src/pubsub.js

```diff
@@ -0,0 +1,586 @@
+                                  .length === 0
+            if (hasAllPeers) {
+              clearInterval(i)
+              callback()
+            }
+          })
+        }, 1000)
+      }
+
+      describe('.publish', () => {
+        it('message from string', (done) => {
+          ipfs1.pubsub.publish(topic, 'hello friend', done)
+        })
+
+        it('message from buffer', (done) => {
+          ipfs1.pubsub.publish(topic, new Buffer('hello friend'), done)
+        })
+      })
+
+      describe('.subscribe', () => {
+        it('to one topic', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+
+            subscription.on('data', (msg) => {
+              expect(msg.data).to.equal('hi')
+              subscription.cancel(done)
+            })
+
+            ipfs1.pubsub.publish(topic, 'hi', (err) => {
+              expect(err).to.not.exist
+            })
+          })
+        })
+
+        it('errors on double subscription', (done) => {
+          series([
+            (cb) => ipfs1.pubsub.subscribe(topic, cb),
+            (cb) => ipfs1.pubsub.subscribe(topic, cb)
+          ], (err, subs) => {
+            expect(err).to.exist
+            expect(err.toString())
+              .to.eql(`Error: Already subscribed to '${topic}'`)
+            subs[0].cancel(done)
+          })
+        })
+
+        it('discover options', (done) => {
+          ipfs1.pubsub.subscribe(topic, {
+            discover: true
+          }, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.cancel(done)
+          })
+        })
+      })
+
+      describe('subscription', () => {
+        it('.cancel and wait for callback', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.cancel(done)
+          })
+        })
+
+        it('.cancel and wait for end event', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.on('end', done)
+            subscription.cancel()
+          })
+        })
+      })
+
+      describe('.peers', () => {
+        before((done) => {
+          ipfs2.id((err, id) => {
+            expect(err).to.not.exist
+            const ipfs2Addr = id.addresses[0]
+            ipfs1.swarm.connect(ipfs2Addr, (err) => {
+              expect(err).to.not.exist
+              // We need to fix this on libp2p level
+              setTimeout(done, 3000)
+            })
+          })
+        })
+
+        // TODO clarify what is the goal of pubsub.peers
+        it('returns an error when not subscribed to a topic', (done) => {
+          ipfs1.pubsub.peers(topic, (err, peers) => {
+            expect(err).to.exist
+            expect(err.toString()).to.equal(`Error: Not subscribed to '${topic}'`)
+            done()
+          })
+        })
```

Comment:
My logic here was that the state of the subscriptions is the source of truth: if you're subscribed, you can query it, cancel it, etc. If you're not, you can't.

What would you expect to receive if you're not subscribed to the topic and call .peers()? En empty list?

Say we have a use case like in the tests for the "waitForPeers", we callback only after we received peers. We don't subscribe to the topic but we keep polling pubsub.peers in the hopes that it returns the peer some day. We keep waiting and waiting and it doesn't seem to work, we check connectivity, we start looking elsewhere, because peers just returns an empty array. That's pretty confusing behaviour for the API. In this case, I believe it's better to tell the user with an error that "hey, what' you're maybe trying to do, will not work".